### PR TITLE
feat(lambda-event-sources): add AT_TIMESTAMP event source mapping starting position

### PR DESCRIPTION
**Pull request by cecheta**
Closes https://github.com/aws/aws-cdk/issues/17214

This PR adds `AT_TIMESTAMP` to the `StartingPosition` enum for lambda event source mappings.

----

### All Submissions:

* [X] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [X] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by cecheta** [comment]
I have updated the integration tests but have not been able to run them.

File: packages/@aws-cdk/aws-lambda-event-sources/lib/kinesis.ts [hunk 1]
```diff
@@ -4,16 +4,24 @@ import * as cdk from '@aws-cdk/core';
 import { StreamEventSource, StreamEventSourceProps } from './stream';
 
 export interface KinesisEventSourceProps extends StreamEventSourceProps {
+  /**
+   * The time from which to start reading, in Unix time seconds.
+   *
+   * @default Required when startingPosition is AT_TIMESTAMP.
+   */
+  readonly startingPositionTimestamp?: number
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-lambda-event-sources/lib/kinesis.ts:12]:
```suggestion
  readonly startingPositionTimestamp?: number;
```

File: packages/@aws-cdk/aws-lambda-event-sources/lib/kinesis.ts [hunk 2]
```diff
@@ -4,16 +4,24 @@ import * as cdk from '@aws-cdk/core';
 import { StreamEventSource, StreamEventSourceProps } from './stream';
 
 export interface KinesisEventSourceProps extends StreamEventSourceProps {
+  /**
+   * The time from which to start reading, in Unix time seconds.
+   *
+   * @default Required when startingPosition is AT_TIMESTAMP.
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-lambda-event-sources/lib/kinesis.ts:10]: This is not the correct use of default. The `@default` tag is meant to describe what happens if left blank -- should be `@default - no timestamp`. The information about when it's required is important, but put that in the regular documentation instead.

File: packages/@aws-cdk/aws-lambda/lib/event-source-mapping.ts [hunk 3]
```diff
@@ -126,6 +126,13 @@ export interface EventSourceMappingOptions {
    */
   readonly startingPosition?: StartingPosition;
 
+  /**
+   * The time from which to start reading, in Unix time seconds.
+   *
+   * @default Required when startingPosition is AT_TIMESTAMP.
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-lambda/lib/event-source-mapping.ts:132]: same comment regarding this `@default`

Comment (kaizencc) [reply packages/@aws-cdk/aws-lambda/lib/event-source-mapping.ts:132]: i also see how `@default` is being used in startingPosition. I consider that wrong also. Would you please change that as well?

File: packages/@aws-cdk/aws-lambda/lib/event-source-mapping.ts [hunk 4]
```diff
@@ -351,4 +362,10 @@ export enum StartingPosition {
    * always read the most recent data in the shard
    */
   LATEST = 'LATEST',
+
+  /**
+   * Start reading from a position defined by a time stamp.
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-lambda/lib/event-source-mapping.ts:367]: This documentation should mention that if supplied, `startingPositionTimestamp` must be set

File: packages/@aws-cdk/aws-lambda/lib/event-source-mapping.ts [hunk 5]
```diff
@@ -299,6 +306,9 @@ export class EventSourceMapping extends cdk.Resource implements IEventSourceMapp
       throw new Error(`tumblingWindow cannot be over 900 seconds, got ${props.tumblingWindow.toSeconds()}`);
     }
 
+    if (props.startingPosition === StartingPosition.AT_TIMESTAMP && !props.startingPositionTimestamp) {
+      throw new Error('startingPositionTimestamp must be provided when startingPosition is AT_TIMESTAMP');
+    }
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-lambda/lib/event-source-mapping.ts:311]: what about when `startingPositionTimestamp` is set but `props.startingPosition !== StartingPosition.AT_TIMESTAMP`? I think we should throw an error also

File: packages/@aws-cdk/aws-lambda/lib/event-source-mapping.ts [hunk 6]
```diff
@@ -351,4 +362,10 @@ export enum StartingPosition {
    * always read the most recent data in the shard
    */
   LATEST = 'LATEST',
+
+  /**
+   * Start reading from a position defined by a time stamp.
+   * Only supported for Amazon Kinesis streams.
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-lambda/lib/event-source-mapping.ts:368]: what happens when this is set for non Kinesis streams?

File: packages/@aws-cdk/aws-lambda-event-sources/lib/kinesis.ts [hunk 7]
```diff
@@ -4,16 +4,24 @@ import * as cdk from '@aws-cdk/core';
...
+   */
+  readonly startingPositionTimestamp?: number
 }
 
 /**
  * Use an Amazon Kinesis stream as an event source for AWS Lambda.
  */
 export class KinesisEventSource extends StreamEventSource {
+  private innerProps: KinesisEventSourceProps;
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-lambda-event-sources/lib/kinesis.ts:19]: I would prefer you not do this and just pull out the props you're going to use. Looks like just a `startingPositionTimestamp` is necessary.

File: packages/@aws-cdk/aws-lambda-event-sources/README.md [hunk 8]
```diff
@@ -196,7 +196,8 @@ behavior:
 * __onFailure__: In the event a record fails and consumes all retries, the record will be sent to SQS queue or SNS topic that is specified here
 * __parallelizationFactor__: The number of batches to concurrently process on each shard.
 * __retryAttempts__: The maximum number of times a record should be retried in the event of failure.
-* __startingPosition__: Will determine where to being consumption, either at the most recent ('LATEST') record or the oldest record ('TRIM_HORIZON'). 'TRIM_HORIZON' will ensure you process all available data, while 'LATEST' will ignore all records that arrived prior to attaching the event source.
+* __startingPosition__: Will determine where to being consumption. 'LATEST' will start at the most recent record and ignore all records that arrived prior to attaching the event source, 'TRIM_HORIZON' will start at the oldest record and ensure you process all available data, while 'AT_TIMESTAMP' will start reading records from a specified time stamp.
+* __startingPositionTimestamp__: The time stamp from which to start reading. Used in conjunction with __startingPosition__.
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-lambda-event-sources/README.md:200]: README needs to mention that `AT_TIMESTAMP` only works for kinesis streams.

**Review by kaizencc — CHANGES_REQUESTED** [review]
Hi @cecheta! Thanks for submitting this PR! I've provided some comments mostly surrounding the documentation around this feature. There's a lot of weird things going on, like the property only being allowed in Kinesis streams and being used in conjunction with `AT_TIMESTAMP`. So we need to be very careful with the docs.

File: packages/@aws-cdk/aws-lambda-event-sources/lib/kinesis.ts (hunk 2 above)
Comment (cecheta) [reply packages/@aws-cdk/aws-lambda-event-sources/lib/kinesis.ts]: Updated

File: packages/@aws-cdk/aws-lambda/lib/event-source-mapping.ts (hunk 3 above)
Comment (cecheta) [reply packages/@aws-cdk/aws-lambda/lib/event-source-mapping.ts]: Updated both

File: packages/@aws-cdk/aws-lambda/lib/event-source-mapping.ts (hunk 4 above)
Comment (cecheta) [reply packages/@aws-cdk/aws-lambda/lib/event-source-mapping.ts:371]: Added

File: packages/@aws-cdk/aws-lambda/lib/event-source-mapping.ts (hunk 5 above)
Comment (cecheta) [reply packages/@aws-cdk/aws-lambda/lib/event-source-mapping.ts:311]: Added another error and a test

File: packages/@aws-cdk/aws-lambda/lib/event-source-mapping.ts (hunk 6 above)
Comment (cecheta) [reply packages/@aws-cdk/aws-lambda/lib/event-source-mapping.ts]: Updated the docs to say that an error will occur

File: packages/@aws-cdk/aws-lambda-event-sources/lib/kinesis.ts (hunk 7 above)
Comment (cecheta) [reply packages/@aws-cdk/aws-lambda-event-sources/lib/kinesis.ts]: Done

File: packages/@aws-cdk/aws-lambda-event-sources/README.md (hunk 8 above)
Comment (cecheta) [reply packages/@aws-cdk/aws-lambda-event-sources/README.md]: Done

File: packages/@aws-cdk/aws-lambda-event-sources/lib/kinesis.ts (hunk 1 above)
Comment (cecheta) [reply packages/@aws-cdk/aws-lambda-event-sources/lib/kinesis.ts]: Fixed

**Review by otaviomacedo — APPROVED** [review]

**Pull request merged**